### DISCUSSION TITLE
Find blank lines in scanned Books

### DISCUSSION THREAD

**Discussion by hollysheep** [ImageMagick/ImageMagick#7125] [Help]
### ImageMagick version

6.9.11

### Operating system, version and so on

Ubuntu

### Description

I have a lot of scanned books and want to convert them into epub. That works really good with FineReader. FR ignores blank lines. Lets say i have this text:

> Text no. 1
> Text no. 2
> 
> Text No. 3

FR would create 3 new paragraphes with the <p>-tag (epub is something like html/css) and doesn't differ between no. 2 and no. 3. I want to find files with that blank lines and correct that manually. all the files are pure text (no imges) and they are trimmed without edges.

Is there an IM script that can do that? It's enough to find the files, i don't need more.

**Comment by fmw42** [comment]
Scale the image down to one column then look for white spaces between dark the size of your line gap using connected components processing.

**Comment by hollysheep** [comment]
Can you give me an example? I don't understand what you mean.

When i try an example from here https://imagemagick.org/script/connected-components.php i have the error that the command 'magick' could not be found. Maybe this is because i've IM 6.x installed? There is no IM 7 in my Ubuntu repo.

**Reply by fmw42** [reply]
For IM 6, use convert in place of magick.

**Reply by fmw42** [reply]
Post and example input image

**Comment by fmw42** [comment]
> all the files are pure text (no images)

Note that Imagemagick is a raster (not text) processor. So your files would have to be rasterized from the text files in order for it to be able to locate and potentially remove the gaps (presumably between paragraphs).

**Comment by fmw42** [comment]
So here is an example.  It first rasterized the text file and scales to 1 column and thresholds to binary. Then it uses connected components to find the bounding boxes for all the white regions that are 10 pixels or more tall in the 1D image. Then it loops over all the bounding boxes and draws rectangles over each of the white regions that it has found.

Input Text File:

[sample3.txt](https://github.com/ImageMagick/ImageMagick/files/14391285/sample3.txt)



```
Lines 1-2: rasterize the text file and save as out.png
Line 3: scale the image to one column and binarize
Lines 4-7: do connected components processing to find the white regions that are 10 or more pixels tall
Lines 8-9: count the number of regions found
```

```
bboxArr=(`convert -size 595x -background white -fill black -pointsize 14 \
caption:"@sample3.txt" +write out.png \
-scale 1x! -negate -threshold 1% -negate -type bilevel \
-define connected-components:verbose=true \
-define connected-components:mean-color=true \
-define connected-components:area-threshold=10- \
-connected-components 8 null: | grep "gray(255)" | awk '{print $2}' | tr "x" "+"`)
num=${#bboxArr[*]}
echo $num
```

```
Line 1: set the right size of the bounding boxes
Line 2: start the looping over the bounding boxes
Lines 3-7: sets the upper left corner and lower right corner of rectangle about each bounding box for the white spaces
Lines 8-10: draw the bounding box rectangles on out.png 
Line 11: end the looping
```

```
ww=$((595-1))
for ((i=0; i<num; i++)); do
bbox=${bboxArr[$i]}
hh=`echo $bbox | cut -d+ -f2`
x1=`echo $bbox | cut -d+ -f3`
y1=`echo $bbox | cut -d+ -f4`
x2=$ww
y2=$((y1+hh-1))
convert out.png -fill none -stroke red -strokewidth 2 \
-draw "rectangle $x1,$y1 $x2,$y2" -alpha off \
out.png
done
```

Result:

![out](https://github.com/ImageMagick/ImageMagick/assets/13458218/5e044bd1-80f8-4ac0-a5f9-b133d1ffd62b)

**Comment by hollysheep** [comment]
I think we have a misunderstanding here. The source file is not a text file its a image created by the scanner (jpg, png or tif). 

I want to search all the images in a folder for files with blank lines. After that i convert the image files into epub using FineReader. IM gives me the names of image files that include a blank line, i search that line in the epub and adjust the p-tag.

**Reply by fmw42** [reply]
I showed above how to find blank lines in a text file that has been scanned.  You just need to take the count of blank lines (num) variable and test if zero or non-zero. Now loop over each file and check the count. Then do whatever else you want with the file that is not related to Imagemagick.
